File: LCA0.java
//LCA1
public TreeNode LCA(TreeNode root, TreeNode one, TreeNode two){
  //base case
  if(root == null) return root;
  if(root == one || root == two) return root;
  //step 1
  TreeNode left = LCA(root.left, one, two);
  TreeNode right = LCA(root.right, one, two);
  //step2+3
  if(left != null && right !== null) return root;
  return left != null ? left : right;
}
//LCA_non-exist
public TreeNode LCA(TreeNode root, TreeNode one, TreeNode two){

}
//LCA_p
public TreeNode LCA(TreeNode root, TreeNode one, TreeNode two){
  int len1 = findLen(root, one);
  int len2 = findLen(root, two);
  int diff = 0;
  if (len1 >= len2){
    diff = len1 - len2;
  }
  while(diff > 0){
    len1--;
    one = one.parent;
  }
  while(one != two){
    one = one.parent;
    two = two.parent;
  }
  return one;
}
private int findLen(TreeNode root, TreeNode node){
  int len = 0;
  while(root != node){
    len++;
    node = node.parent;
  }
  return len;
}
//LCA_k_node
public TreeNode LCA(TreeNode root, List<TreeNodes> list){
  Set<TreeNode> set = new HashSet<>();
  for(TreeNode each : list){
    set.add(each);
  }
  return helper(root, set);
}
private TreeNode helper(TreeNode root, Set<TreeNode> set){
  if(root == null) return root;
  if(set.contains(root)) return root;
  //step 1
  TreeNode left = LCA(root.left, set);
  TreeNode right = LCA(root.right, set);
  //step2+3
  if(left != null && right !== null) return root;
  return left != null ? left : right;
}
//LCA_k_nary
public TreeNode LCA(TreeNode root, TreeNode one, TreeNode two){
  //base case
  if(root == null) return root;
  if(root == one || root == two) return root;
  //step 1
  int counter = 0;
  TreeNode temp = null;
  for(TreeNode each : children){
    TreeNode node = helper(each, one, two);
    if(node != null){
      counter++;
      if(counter > 1) return root;
      temp = node;
    }
  }
  return temp;
}
//LCA_kk
public TreeNode LCA(TreeNode root, List<TreeNodes> list){
  Set<TreeNode> set = new HashSet<>();
  for(TreeNode each : list){
    set.add(each);
  }
  return helper(root, set);
}
private TreeNode helper(TreeNode root, Set<TreeNode> set){
  if(root == null) return root;
  if(set.contains(root)) return root;
  //step 1
  int counter = 0;
  TreeNode temp = null;
  for(TreeNode each : children){
    TreeNode node = helper(each, set);
    if(node != null){
      counter++;
      if(counter > 1) return root;
      temp = node;
    }
  }
  return temp;
}
//LCA_bst
public TreeNode LCA(TreeNode root, TreeNode one, TreeNode two){
  if(root == null) return root;
  if(one.key < root.key && two.key < root.key){
    return LCA(root.left, one, two);
  }else if(one.key > root.key && two.key > root.key){
    return LCA(root.right, one, two);
  }else{
    return root;
  }
}

File: tree/LCA/LCA0.py
#LCA1   O(height)
def LCA(root, one, two):
    if root is None:
        return root
    if root is one or root is two:
        return root
    left = LCA(root.left, one, two)
    right = LCA(root.right, one, two)
    if left and right:
        return root
    else:
        return left is left else right

#LCA_non-exist
def non_LCA(root, one, two):
    return root
#LCA_p    O(n)
def LCA(root, one, two):
    if root is None:
        return root
    if root is one or root is two:
        return root
    one_dist = getLen(root, one)
    two_dist = getLen(root, two)
    if one_dist > two_dist:
        for i in range(one_dist - two_dist):
            one = one.parent
    else:
        for i in range(two_dist - one.dist):
            two = two.parent
    while one != two:
        one = one.parent
        two = two.parnet
    return one

def getLen(root, node):
    res = 0
    cur = node
    while cur != node
        res += 1
        cur = cur.parent
    return res

#LCA_k_node
def LCA_knode(root, knodes_set):
    if root is None:
        return root
    if root in knodes_set:
        return root
    left = LCA_knode(root.left, knodes_set)
    right = LCA_knode(root.right, knodes_set)
    if left and right:
        return root
    else
        return left if left else right

#LCA_k_nary
# node.children
# need to check more than 1 child return not null
def LCA_knary(root, one, two):
    if root is None:
        return root
    if root is one or root is two:
        return root
    temp = None
    counter = 0
    for child in root.children:
        node = LCA_knary(child, one, two)
            if node:
                counter += 1
                if counter == 1:
                    return root
            else:
                temp = node
    return temp

#LCA_kk
# think should be the same as above
def LCA_kk(root, knodes_set):
    if root is None:
        return root
    if root is in knodes_set:
        return root
    counter = 0
    temp = None
    for child in root.children:
        node = LCA_kk(child, knodes_set)
        if node:
            counter += 1
            if counter == 2:
                return root
            else:
                temp = node
    return temp

#LCA_bst
def LCA_bst(root, one, two):
    if root is None:
        return root
    if root.val < one.val and root.val < two.val:
        return LCA_bst(root.right, one, two)
    elif root.val > one.val and root.val > two.val:
        return LCA_bst(root.left, one, two)
    else:
        reutrn root
